### PR TITLE
Add Python 3.12 to CI now that p4p is updated

### PR DESCRIPTION
**Pull request by OCopping**
I noticed Python 3.12 never got added to the CI after support for it in `p4p` got released.

### CODE REVIEW

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]
If we're supporting 3.12 add it to the pyproject.toml too
Good work getting this over the line :partying_face:

**Comment by OCopping** [comment]
The tests are failing because of a `get_event_loop()` deprecation warning/error.
From the [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop):
```Deprecated since version 3.12: Deprecation warning is emitted if there is no current event loop. In some future Python release this will become an error.```

**Comment by coretl** [comment]
> The tests are failing because of a `get_event_loop()` deprecation warning/error. From the [docs](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop): `Deprecated since version 3.12: Deprecation warning is emitted if there is no current event loop. In some future Python release this will become an error.`

@evalott100 I think this is coming from the "detect pending tasks" fixture, would you mind taking a look please?

**Comment by evalott100** [comment]
Taking this over.

File: tests/epics/adcore/test_drivers.py [hunk 1]
```diff
@@ -90,10 +90,12 @@ async def wait_then_fail():
...
+
     acquiring = await adcore.start_acquiring_driver_and_ensure_status(
         driver, timeout=0.1
     )
-    await wait_then_fail()
-
-    with pytest.raises(ValueError):
+    with pytest.raises(
+        ValueError, match="Final detector state DetectorState.DISCONNECTED"
+    ):
```
Comment (evalott100) [inline tests/epics/adcore/test_drivers.py:100]: We should be starting setting up `wait_then_fail` before we schedule the task which should fail.

File: panda_connect.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        "PANDA1:", StaticPathProvider(StaticFilenameProvider("test-panda"), Path("."))
+    )
+    await panda.connect()
+    assert isinstance(panda, HDFPanda)
+    table = await panda.seq[1].table.get_value()
+    assert isinstance(table, SeqTable)
+    print("OUTA1", table.outa1)
+    outa2 = table.outa2
+    position = table.position
+    assert isinstance(outa2, np.ndarray) and outa2.dtype == np.bool_, outa2.dtype
```
Comment (DiamondJoseph) [inline panda_connect.py]:
```suggestion
    assert isinstance(outa2, np.ndarray) and outa2.dtype == np.bool_, f"{outa2.dtype} is of wrong type!"
```

Not exactly the wording I want, but I was confused until I realised the arg after the comma was for the AssertionError, not part of the comparison

File: panda_connect.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    assert (
+        isinstance(position, np.ndarray) and position.dtype == np.int32
+    ), position.dtype
```
Comment (DiamondJoseph) [inline panda_connect.py]: as above

File: panda_connect.py (hunk 2 above)
Comment (DiamondJoseph) [reply panda_connect.py]: But is this for local testing only and not part of the code to be pushed?

File: tests/conftest.py [hunk 4]
```diff
@@ -113,15 +115,21 @@ def fail_test_on_unclosed_tasks(request: FixtureRequest):
     by the end of the test.
     """
 
-    fail_count = request.session.testsfailed
-    loop = asyncio.get_event_loop()
-    loop.set_debug(True)
+    try:
+        fail_count = request.session.testsfailed
+        loop = asyncio.get_running_loop()
```
Comment (evalott100) [inline tests/conftest.py:120]: `get_event_loop` gives a deprecation warning that future python versions will raise an error if there is no event loop, `get_running_loop` does this already. It doesn't seem possible to check for an event loop without failing/warning if there isn't one, so we just catch the exception for the tests which don't use one.

File: tests/conftest.py [hunk 5]
```diff
@@ -113,15 +115,21 @@ def fail_test_on_unclosed_tasks(request: FixtureRequest):
...
-            loop, request.node.name, request.session.testsfailed == fail_count
+        request.addfinalizer(
+            lambda: _error_and_kill_pending_tasks(
+                loop, request.node.name, request.session.testsfailed == fail_count
+            )
         )
-    )
+    except RuntimeError as error:
+        if str(error) != "no running event loop":
+            raise error
```
Comment (DiamondJoseph) [inline tests/conftest.py:132]: nit: Is it worth having a `pass` explaining that if `str(error) == "no running event loop"` then the test either does not require a loop or will fail when attempting to use the loop?

**Review by DiamondJoseph — APPROVED** [review]
Looks good from my side, but as I'm not as involved in ophyd-async consider this advisory only

File: tests/conftest.py (hunk 5 above)
Comment (evalott100) [reply tests/conftest.py:131]: > Is it worth having a `pass`

What do you mean by this?
``` python
except RuntimeError as error:
    if str(error) == "no running event loop":
        pass
    else:
        raise error
```
?

Comment (DiamondJoseph) [reply tests/conftest.py:131]: Yes exactly that, just a bit more obvious what is intended

Comment (evalott100) [reply tests/conftest.py:131]: I would prefer the current `!=` comprehension but I'll add a comment explaining!

File: tests/core/test_observe.py [hunk 6]
```diff
@@ -86,7 +86,7 @@ async def tick():
 
     async def watch():
         async for val in observe_value(sig, done_timeout=0.2):
-            time.sleep(0.15)
+            await asyncio.sleep(0.1)
```
Comment (coretl) [inline tests/core/test_observe.py:89]: Ok, my fault for not putting a comment in here, but this really was meant to be a `time.sleep`! This is a test to prove a subtle timing bug where the inner loop of observe_value was blocking the event loop. Please can you restore this and put a comment to this effect in?

Comment (evalott100) [reply tests/core/test_observe.py:89]: Made a [new issue](https://github.com/bluesky/ophyd-async/pull/692), easier than a revert.

**Pull request merged**